File: DSA_Striver/Day11/Heaps/MergeKSortedLists.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for singly-linked list
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Compare {
public:
    // Comparator to order ListNode pointers based on node values
    bool operator()(ListNode* a, ListNode* b) {
        return a->val > b->val;
    }
};

class Solution {
public:
    // Function to merge k sorted linked lists using a min-heap
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // Create a min-heap (priority queue) with custom comparator
        priority_queue<ListNode*, vector<ListNode*>, Compare> pq;

        // Push the head of each non-empty list into the heap
        for (auto list : lists) {
            if (list != NULL)
                pq.push(list);
        }

        // Create a dummy node to build the result list
        ListNode* dummy = new ListNode(0);
        ListNode* tail = dummy;

        // While the heap is not empty
        while (!pq.empty()) {
            // Extract the node with the smallest value
            ListNode* smallest = pq.top();
            pq.pop();

            // Add it to the result list
            tail->next = smallest;
            tail = tail->next;

            // If there's a next node, push it into the heap
            if (smallest->next != NULL)
                pq.push(smallest->next);
        }

        // Return the head of the merged list
        return dummy->next;
    }
};

// Driver code
int main() {
    Solution sol;

    // Creating three linked lists:
    // list1: 1 -> 4 -> 5
    // list2: 1 -> 3 -> 4
    // list3: 2 -> 6

    ListNode* list1 = new ListNode(1);
    list1->next = new ListNode(4);
    list1->next->next = new ListNode(5);

    ListNode* list2 = new ListNode(1);
    list2->next = new ListNode(3);
    list2->next->next = new ListNode(4);

    ListNode* list3 = new ListNode(2);
    list3->next = new ListNode(6);

    vector<ListNode*> lists = {list1, list2, list3};
    ListNode* result = sol.mergeKLists(lists);

    // Print the merged list
    while (result != NULL) {
        cout << result->val << " ";
        result = result->next;
    }
    return 0;
}

File: c++/Heaps/MergeKsortedLists.java
import java.util.*;

class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
    }
}

class Solution {

    public ListNode mergeKLists(ListNode[] lists) {

        // Min-heap: smallest node comes first
        PriorityQueue<ListNode> pq =
            new PriorityQueue<>((a, b) -> a.val - b.val);

        // Put first node of every list into heap
        for (ListNode head : lists) {
            if (head != null) {
                pq.add(head);
            }
        }

        // Dummy node
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;

        while (!pq.isEmpty()) {

            // Get smallest node
            ListNode curr = pq.poll();

            // Add it to answer
            tail.next = curr;
            tail = curr;

            // Add next node from same list
            if (curr.next != null) {
                pq.add(curr.next);
            }
        }

        return dummy.next;
    }
}